addon/debounce-task.js

```diff
@@ -0,0 +1,148 @@
+
+   export default Component.extend({
+     logMe() {
+       console.log('This will only run once every 300ms.');
+     },
+
+     click() {
+       debounceTask(this, 'logMe', 300);
+     },
+
+     destroy() {
+       runDisposables(this);
+     }
+   });
+   ```
+
+   @method debounceTask
+   @param { Object } obj the instance to register the task for
+   @param { String } name the name of the task to debounce
+   @param { ...* } debounceArgs arguments to pass to the debounced method
+   @param { Number } wait the amount of time to wait before calling the method (in milliseconds)
+   @public
+   */
+export function debounceTask(obj, name, ...debounceArgs) {
+  assert(
+    `Called \`debounceTask\` without a string as the first argument on ${obj}.`,
+    typeof name === 'string'
+  );
+  assert(
+    `Called \`obj.debounceTask('${name}', ...)\` where 'obj.${name}' is not a function.`,
+    typeof obj[name] === 'function'
+  );
+  assert(
+    `Called \`debounceTask\` on destroyed object: ${obj}.`,
+    !obj.isDestroyed
+  );
+
+  let pendingDebounces = getOrAllocate(
+    registeredDebounces,
+    obj,
+    Object,
+    getDebouncesDisposable
+  );
+  let debounce = pendingDebounces[name];
+  let debouncedTask;
+
+  if (!debounce) {
+    debouncedTask = (...args) => {
+      delete pendingDebounces[name];
+      obj[name](...args);
+    };
+  } else {
+    debouncedTask = debounce.debouncedTask;
+  }
+
+  // cancelId is new, even if the debounced function was already present
+  let cancelId = run.debounce(obj, debouncedTask, ...debounceArgs);
+
+  pendingDebounces[name] = { debouncedTask, cancelId };
+}
+
+/**
+   Cancel a previously debounced task.
+
+   Example:
+
+   ```js
+   import Component from 'ember-component';
+   import { debounceTask, cancelDebounce } from 'ember-lifeline';
+
+   export default Component.extend({
+     logMe() {
+       console.log('This will only run once every 300ms.');
+     },
+
+     click() {
+       debounceTask(this, 'logMe', 300);
+     },
+
+     disable() {
+        cancelDebounce(this, 'logMe');
+     },
+
+     destroy() {
+       runDisposables(this);
+     }
+   });
+   ```
+
+   @method cancelDebounce
+   @param { Object } obj the instance to register the task for
+   @param { String } methodName the name of the debounced method to cancel
+   @public
+   */
+export function cancelDebounce(obj, name) {
+  let pendingDebounces = registeredDebounces.get(obj);
+
+  if (pendingDebounces === undefined) {
+    return;
+  }
+
+  let { cancelId } = pendingDebounces[name];
+
+  run.cancel(cancelId);
+}
+
+function getDebouncesDisposable(debounces) {
+  return function() {
+    let debounceNames = debounces && Object.keys(debounces);
+
+    if (!debounceNames || !debounceNames.length) {
+      return;
+    }
+
+    for (let i = 0; i < debounceNames.length; i++) {
+      let { cancelId } = debounces[debounceNames[i]];
+
+      cancelTask(cancelId);
```

Comment:
Seems odd that this goes through `cancelTask` but the `cancelDebounce` method above uses `run.cancel(cancelId)` directly.  I would have expected them both to go through the same path...

Reply:
I'm fine with explicitly calling `run.cancel` here to. Makes no diff to me.

Reply:
Fixed.